File: src/Luke04/Program.cs
using System;
using System.IO;
using System.Linq;

namespace Luke04
{
	class Program
	{
		static void Main(string[] args)
		{
			var totalIngredients = File.ReadAllLines("leveringsliste.txt")
				.Select(ParseLine)
				.Aggregate(new Delivery(), (acc, curr) =>
				{
					acc.Sugar += curr.Sugar;
					acc.Flour += curr.Flour;
					acc.Milk += curr.Milk;
					acc.Eggs += curr.Eggs;
					return acc;
				});

			var count = 0;
			while (true)
			{
				if (totalIngredients.Sugar >= 2 &&
				    totalIngredients.Flour >= 3 &&
				    totalIngredients.Milk >= 3 &&
				    totalIngredients.Eggs >= 1)
				{
					count++;
					
					totalIngredients.Sugar -= 2;
				    totalIngredients.Flour -= 3;
				    totalIngredients.Milk -= 3;
				    totalIngredients.Eggs -= 1;
				}
				else
				{
					break;
				}
			}
			
			Console.WriteLine(count);
		}

		private static Delivery ParseLine(string line) =>
			line.Split(',')
				.Aggregate(new Delivery(), (acc, curr) =>
				{
					var split = curr.Split(':').Select(s => s.Trim(' ', ',')).ToList();
					switch (split[0])
					{
						case "sukker":
							acc.Sugar = int.Parse(split[1]);
							break;
						case "mel":
							acc.Flour = int.Parse(split[1]);
							break;
						case "melk":
							acc.Milk = int.Parse(split[1]);
							break;
						case "egg":
							acc.Eggs = int.Parse(split[1]);
							break;
					}

					return acc;
				});
	}

	class Delivery
	{
		public int Sugar { get; set; }
		public int Flour { get; set; }
		public int Milk { get; set; }
		public int Eggs { get; set; }
	}
}

File: src/fsharp/Luke04/Program.fs
open System
open System.IO

type Delivery = {
    Sugar: int
    Flour: int
    Milk: int
    Eggs: int
}

let sufficientQuantities delivery =
    delivery.Sugar >= 2 && delivery.Flour >= 3 && delivery.Milk >= 3 && delivery.Eggs >= 1

[<EntryPoint>]
let main argv =
    File.ReadAllLines "leveringsliste.txt"
    |> Array.toList
    |> List.collect (fun line -> line.Split(',') |> Array.toList)
    |> List.map (fun pair -> pair.Split(':')
                             |> Array.toList
                             |> List.map (fun s -> s.Trim(' ', ',')))
    |> List.fold (fun total pair ->
                  match pair with
                  | i::q::_ when i = "sukker" -> {total with Sugar = total.Sugar + int q}
                  | i::q::_ when i = "mel" -> {total with Flour = total.Flour + int q}
                  | i::q::_ when i = "melk" -> {total with Milk = total.Milk + int q}
                  | i::q::_ when i = "egg" -> {total with Eggs = total.Eggs + int q}
                  | _ -> total) { Sugar = 0; Flour = 0; Milk = 0; Eggs = 0 }
    |> List.unfold (fun total -> if sufficientQuantities total
                                 then Some(1, { Sugar = total.Sugar - 2; Flour = total.Flour - 3; Milk = total.Milk - 3; Eggs = total.Eggs - 1 })
                                 else None)
    |> List.sum
    |> printfn "%i"
    
    0 // return an integer exit code
